### PR TITLE
[TableGen] Add `PreferSmallerInstructions` for Targets.

### PR DESCRIPTION
**Pull request by AlfieRichardsArm**
This option means that in assembly matching instructions with smaller encodings will be preferred.

This will be used for the ARM instruction set where this is the correct behavior after some other refactoring.

### CODE REVIEW

File: llvm/include/llvm/Target/Target.td
Comment (kosarev) [inline llvm/include/llvm/Target/Target.td:1728]: Typo: `=` -> `-`

**Review by kosarev — APPROVED** [review]
Thanks. LGTM modulo the typo. But please don't submit until this is actually needed.

File: llvm/include/llvm/Target/Target.td
Comment (AlfieRichardsArm) [reply llvm/include/llvm/Target/Target.td]: Done

**Comment by AlfieRichardsArm** [comment]
@s-barannikov Regarding 
> I think this check should be tie-breaking rather than defining, and so it should be moved below, after we've compared the number of operands and their classes.
> Also, does this check supersede the AVX/AVX512 resolution below?

It is my understanding this needs to be here for ARM. The architecture reference manual states: 
```
If neither .W nor .N is specified, the assembler can select either a 16-bit or 32-bit encoding. If both encoding
lengths are available, it must select a 16-bit encoding.
```
There are cases (such as thumb 1 instructions with a CCOut operand whereas thumb 2 doesnt) where the smaller instruction has more operands, so choosing that first would make this non-compliant.

**Comment by AlfieRichardsArm** [comment]
@s-barannikov regarding
> Should it be a member of AsmParser instead?
I have just pushed a commit that moves it.

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp
Comment (s-barannikov) [inline llvm/utils/TableGen/AsmMatcherEmitter.cpp:676]: The comment needs to be fixed. This function does not sort things.

**Review by s-barannikov — APPROVED** [review]

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp
Comment (AlfieRichardsArm) [reply llvm/utils/TableGen/AsmMatcherEmitter.cpp:676]: Done

Comment (AlfieRichardsArm) [reply llvm/utils/TableGen/AsmMatcherEmitter.cpp:676]: Done, thank you for catching.

File: llvm/utils/TableGen/CodeGenTarget.cpp
Comment (AlfieRichardsArm) [reply llvm/utils/TableGen/CodeGenTarget.cpp:345]: Sure!

File: llvm/utils/TableGen/AsmMatcherEmitter.cpp
Comment (AlfieRichardsArm) [reply llvm/utils/TableGen/AsmMatcherEmitter.cpp:3255]: Done

**Comment by AlfieRichardsArm** [comment]
```CPP
-        if (A.couldMatchAmbiguouslyWith(B, Target)) {
+        bool PreferSmallerInstructions =
+            AsmParser->getValueAsBit("PreferSmallerInstructions");
+        if (A.couldMatchAmbiguouslyWith(B, PreferSmallerInstructions)) {
```

> @s-barannikov: Same for shouldBeMatchedBefore above.

Fixed.

**Comment by AlfieRichardsArm** [comment]
https://github.com/llvm/llvm-project/pull/83436#discussion_r1524596531
> I would add to this comment
> 
> (a) that this flag is useful for Arm (so that people who only know other architectures have an idea why it needs to be here in the first place)

Adding this now

> (b) what the preference order is if this flag is false. (Surely the only reason this flag exists, instead of just being always-true, is so that somebody can deliberately set it to false, in which case they have some reason to want a different preference order, and will need to know how to get it.)

The main reason for this flag is this change breaks a lot of assumptions that other backends (MIPS primarily) rely upon (see all the assumptions it broke in the ARM backend to get an idea (: ). I will add a comment to explain the default ordering and the change this makes.

**Review by statham-arm — APPROVED** [review]
I'm happy with this from the point of view of it filling in the missing comment I asked for on review #83436. I can't comment on whether the details of the specified preference order are _accurate_, though, so I have to rely on other reviewers for that :-)

**Pull request merged**
